sources/dma_driver.c

```diff
@@ -0,0 +1,719 @@
+ *          read/written. These bits can only be written when the stream is disabled. This is 
+ *          the first of two memory addresses available. THe second address is only used for 
+ *          double buffer mode which is not implemented in this driver. 
+ * 
+ * @param dma_stream : pointer to DMA port stream being configured 
+ * @param m0ar : memory 0 address 
+ */
+void dma_m0ar(
+    DMA_Stream_TypeDef *dma_stream, 
+    uint32_t m0ar); 
+
+
+/**
+ * @brief FIFO error interrupt 
+ * 
+ * @see dma_feie_t
+ * 
+ * @param dma_stream : pointer to DMA port stream being configured 
+ * @param feie : FIFO error interrupt configuration 
+ */
+void dma_feie(
+    DMA_Stream_TypeDef *dma_stream, 
+    dma_feie_t feie); 
+
+
+/**
+ * @brief Direct/FIFO mode selection 
+ * 
+ * @details Configures FIFO mode. If FIFO mode is enabled then the FIFO buffer will be filled 
+ *          with data from the source by DMA transfers up to the defined FIFO buffer threshold 
+ *          at which point the FIFO will be drained to the destination. If not enabled, the 
+ *          stream will be in direct mode where the FIFO buffer is drained immediately after 
+ *          a data item enters it. <br> 
+ *          
+ *          If memory-to-memory mode is selected and the stream is enabled then FIFO mode is 
+ *          automatically enabled as direct mode is not allowed in this circumstance. This mode 
+ *          can only be written when the stream is disabled. 
+ * 
+ * @see dma_fifo_mode_t
+ * @see dma_fth
+ * 
+ * @param dma_stream : pointer to DMA port stream being configured 
+ * @param mode : FIFO mode configuration 
+ */
+void dma_dmdis(
+    DMA_Stream_TypeDef *dma_stream, 
+    dma_fifo_mode_t mode); 
+
+
+/**
+ * @brief FIFO threshold selection 
+ * 
+ * @details Configures the threshold of the FIFO buffer. When the buffer reaches this 
+ *          threshold then the data within it gets drained. This functionality is only used 
+ *          in FIFO mode. These bits can only be written when the stream is disabled. 
+ * 
+ * @see dma_fifo_threshold_t
+ * @see dma_dmdis
+ * 
+ * @param dma_stream : pointer to DMA port stream being configured 
+ * @param fth : FIFO threshold configuration 
+ */
+void dma_fth(
+    DMA_Stream_TypeDef *dma_stream, 
+    dma_fifo_threshold_t fth); 
+
+//================================================================================
+
+
+//================================================================================
+// Initialization 
+
+// Initialize the DMA stream 
+void dma_stream_init(
+    DMA_TypeDef *dma, 
+    DMA_Stream_TypeDef *dma_stream, 
+    dma_channel_t channel, 
+    dma_direction_t dir, 
+    dma_cm_t cm, 
+    dma_priority_t priority, 
+    dma_addr_inc_mode_t minc, 
+    dma_addr_inc_mode_t pinc, 
+    dma_data_size_t msize, 
+    dma_data_size_t psize)
+{
+    // Enable the DMA clock
+    if (dma == DMA1)
```

Comment:
Nit:

Typically avoid using the no-brace if/else shortcuts.

They can lead to errors/confusion if other developers modify without noticing